### PR TITLE
[TRY] Core data: extend stable key to use per_page and page query parts

### LINKED ISSUE #56355: Core data: selectors return stale data despite different query parameters
### Description

Unresolved core data selectors will return `null` if records are not found in state. They'll then go off and "resolve" the query (fetch the data from the API) and update the state.

When using query params such as `per_page`  — params that normally should trigger a new API call given that it can request records not yet available in the state — the selectors will not return `null` but return whatever it finds in the state until the resolvers are done.

From what I can gather, the `queries` reducer state for each entity [creates "stableKeys"](https://github.com/WordPress/gutenberg/blob/d4c367d25a87d46e26183f5958d384f350279b28/packages/core-data/src/queried-data/get-query-parts.js#L100-L100) to cache IDs for certain queries. Currently only `_fields` and `includes` queries are cached, otherwise the key is `""`. 

Props to @andrewserong for finding and reporting here: https://github.com/WordPress/gutenberg/pull/54046#pullrequestreview-1739021045

### Step-by-step reproduction instructions

Make sure you have a few posts or pages or other entities saved.

In your browser's console, fetch those entities:

```js
 // 1. API call, returns null
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 2 } )

 // 2. No API call, returns expected results
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 2 } )

 // 3.  API call, returns results from previous command 2
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 3 } )

 // 4.  No API call, returns expected results
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 3 } )

 // 5.  API call, returns results from previous command 3
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 4 } )

 // 6.  No API call, returns expected results
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 4 } ) 
```

### What I expected
For commands 3 and 4, we'd expect either `null` or `[]` or some other result to indicate that the selector cannot find these records.

Instead it returns the last known query.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

- WordPress 6.5 alpha
- Gutenberg latest (trunk)

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ramonjd**
Maybe resolves https://github.com/WordPress/gutenberg/issues/56355

Follow up to https://github.com/WordPress/gutenberg/pull/54046

❗ I'm not sure this is the right way forward. There's [a test that explicitly checks that stableKey does not contain any pagination data](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/queried-data/test/get-query-parts.js#L7). I'm not sure why.

------- 

## What?
This is a test to check whether extending the stable key with params, which should trigger new API calls, can be added to the state: `per_page`.

## How?
Adding an entry to the `queries` reducer every time the `per_page` query part changes

Screenshot 2023-11-21 at 3 08 45 pm

## Testing Instructions

Does this break anything?

```
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 2 } ) // API call, returns null
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 2 } ) // No API call, returns results

await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 3 } ) // API call, returns null
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 3 } ) // No API call, returns results

await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 4 } ) // API call, returns null
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 4 } )  // No API call, returns results
```

### CODE REVIEW

File: packages/core-data/src/queried-data/test/get-query-parts.js [hunk 1]
```diff
@@ -4,14 +4,14 @@
 import { getQueryParts } from '../get-query-parts';
 
 describe( 'getQueryParts', () => {
-	it( 'parses out pagination data', () => {
+	it( 'parses out pagination data and adds to returned object', () => {
```
Comment (ramonjd) [inline packages/core-data/src/queried-data/test/get-query-parts.js:7]: Just updating these tests due to OCD. 

Given they they're checking for empty stableKeys here, I'm not very sure about this approach.

**Review by youknowriad — COMMENTED** [review]
Ok, this looks correct to me but I'm curious about impact as well.

**Comment by ntsekouras** [comment]
I recently encountered this issue in a different [use case](https://github.com/WordPress/gutenberg/pull/55316). There's a lot of complexity there, but I think what we do is handle the `per_page` to keep the cached results [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/queried-data/reducer.js#L40) and this makes sense to me, because in `getQueriedItemsUncached` we used the cached value.

What this means is if we have a query with `perPage: 5` for a specific query, and we trigger the exact same query with `perPage:2`, we can just take it from cache.

From the issue description, it's a valid issue, though if we use the selector with a subscription to the store, the updated value is used later. So I'm not sure yet how we should proceed, but if we want to add `per_page` in the stable key, we should more changes, as it makes other parts of code unused(ex. getQueriedItemsUncached).

**Comment by youknowriad** [comment]
What happens when you have both a page value other than 1 and a per page. In this case, using the stale is clearly a bug because on the second page, items won't start at the same position

**Comment by ramonjd** [comment]
**TL;DR** This PR disrupts existing pagination caching. I'll close this and move the discussion to the [issue](https://github.com/WordPress/gutenberg/issues/56355).


> There's a lot of complexity there, but I think what we do is handle the per_page to keep the cached results [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/queried-data/reducer.js#L40) and this makes sense to me, because in getQueriedItemsUncached we used the cached value.

Thanks for the helpful context. 

Yeah, there's a lot going on 😄 

As @ntsekouras points out, `getMergedItemIds` tracks loaded ids and their paginated position, by maintaining an array of "paginated" results in the correct order, creating `undefined` entries where the item has not yet been fetched, e.g., `{page:2, per_page:2} === [undefined, undefined, 1, 2]`.

I've tried to illustrate this over in a code sandbox: https://codesandbox.io/p/sandbox/cool-sunset-gyfxgy

Core data maintains this state in the queries reducer under an empty key `""`.

Screenshot 2023-11-22 at 1 37 10 pm

_I don't think this PR is the right way forward_ as it kinda breaks all this - it's adding `"per_page=n"` keys to the queries reducer state so we can't look up cached "paginated" results.

Example queries on this PR:

```js
// Fetches the first 3 records from API.
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 3 } );

// Fetches the first 3 records from state (cached).
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 3 } )

// Fetches the first 2 records from API, but returns null << we've already got these so null is wrong
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 2 } )
```

> What this means is if we have a query with perPage: 5 for a specific query, and we trigger the exact same query with perPage:2, we can just take it from cache.

On trunk, I'm still seeing an API call even when we've cached results, is that right?

```js
// Fetches the first 5 records from API.
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 5 } );

// Fetches the first 5 records from state (cached).
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 5 } )

// Fetches the first 2 records from API, AND fetches the first 2 records from API << we've already got these and 
// the cache hasn't been invalidated, why make an API call?
await wp.data.select( 'core' ).getEntityRecords( 'postType', 'post', { per_page: 2 } )
```

So yeah, maybe something is going on with the way we're checking the cache for pagination. 🤔  I'd expect something along the lines of the following:

1. Are there corresponding items in the queries array? E.g., `[undefined, undefined, 1, 2]`
2. If not, fetch from API and return `null`
3. If yes and resolver hasn't been invalidated, grab out of cache and return results

**Comment by ramonjd** [comment]
I'll keep thinking about this 🤔

**Comment by andrewserong** [comment]
I don't have anything else to add, but great discussion here folks, and thanks for digging into the issue @ramonjd! While this PR wasn't the right one, it's helped hone in on a desired path forwards that can re-use cached items 👍
